libkbfs/backpressure_disk_limiter.go

```diff
@@ -239,6 +240,310 @@ func (bt *backpressureTracker) getStatus() backpressureTrackerStatus {
+	quotaBytes int64
+}
+
+func newQuotaBackpressureTracker(minThreshold, maxThreshold float64) (
+	*quotaBackpressureTracker, error) {
+	if minThreshold < 0.0 {
+		return nil, errors.Errorf("minThreshold=%f < 0.0",
+			minThreshold)
+	}
+	if maxThreshold < minThreshold {
+		return nil, errors.Errorf(
+			"maxThreshold=%f < minThreshold=%f",
+			maxThreshold, minThreshold)
+	}
+	qbt := &quotaBackpressureTracker{
+		minThreshold, maxThreshold, 0, 0, math.MaxInt64,
+	}
+	return qbt, nil
+}
+
+func (qbt quotaBackpressureTracker) usedFrac() float64 {
+	return (float64(qbt.unflushedBytes) + float64(qbt.remoteUsedBytes)) /
+		float64(qbt.quotaBytes)
+}
+
+// delayScale returns a number between 0 and 1, which should be
+// multiplied with the maximum delay to get the backpressure delay to
+// apply.
+func (qbt quotaBackpressureTracker) delayScale() float64 {
+	usedFrac := qbt.usedFrac()
+
+	// We want the delay to be 0 if usedFrac <= m and the max
+	// delay if usedFrac >= M, so linearly interpolate the delay
+	// scale.
+	//
+	// TODO: Consider a scheme which applies backpressure
+	// differently if R/Q > m, but (U+R)/Q <= 1. In that case, we
+	// don't want to "permanently" apply backpressure before the
+	// quota is actually hit. In that case, the backpressure
+	// should probably be proportional to U/(Q-R) until (U+R)/Q =
+	// 1, at which point it should smoothly transition to a
+	// formula that is proportional to (U+R)/Q.
+	m := qbt.minThreshold
+	M := qbt.maxThreshold
+	return math.Min(1.0, math.Max(0.0, (usedFrac-m)/(M-m)))
+}
+
+func (qbt *quotaBackpressureTracker) onJournalEnable(unflushedBytes int64) {
+	qbt.unflushedBytes += unflushedBytes
+}
+
+func (qbt *quotaBackpressureTracker) onJournalDisable(unflushedBytes int64) {
+	qbt.unflushedBytes -= unflushedBytes
+}
+
+func (qbt *quotaBackpressureTracker) updateRemote(
+	remoteUsedBytes, quotaBytes int64) {
+	qbt.remoteUsedBytes = remoteUsedBytes
+	qbt.quotaBytes = quotaBytes
+}
+
+func (qbt *quotaBackpressureTracker) afterBlockPut(
+	blockBytes int64, putData bool) {
+	if putData {
+		qbt.unflushedBytes += blockBytes
+	}
+}
+
+func (qbt *quotaBackpressureTracker) onBlocksFlush(blockBytes int64) {
+	qbt.unflushedBytes -= blockBytes
+}
+
+type quotaBackpressureTrackerStatus struct {
+	// Derived numbers.
+	UsedFrac   float64
+	DelayScale float64
+
+	// Constants.
+	MinThreshold float64
+	MaxThreshold float64
+
+	// Raw numbers.
+	UnflushedBytes  int64
+	RemoteUsedBytes int64
+	QuotaBytes      int64
+}
+
+func (qbt *quotaBackpressureTracker) getStatus() quotaBackpressureTrackerStatus {
+	return quotaBackpressureTrackerStatus{
+		UsedFrac:   qbt.usedFrac(),
+		DelayScale: qbt.delayScale(),
+
+		MinThreshold: qbt.minThreshold,
+		MaxThreshold: qbt.maxThreshold,
+
+		UnflushedBytes:  qbt.unflushedBytes,
+		RemoteUsedBytes: qbt.remoteUsedBytes,
+		QuotaBytes:      qbt.quotaBytes,
+	}
+}
+
+// journalTracker aggregates all the journal trackers.
+type journalTracker struct {
```

Comment:
Bundled all the journal-related trackers together, as to move towards a world where it's less coupled from the disk cache tracker.